### ISSUE TITLE
How to freeze batchnorm parameter when using AffineChannel layer?

### ISSUE THREAD

**Issue by lsrock1** [facebookresearch/Detectron#586]
I am trying to build detectron with my custom network backbone.

1.
I am using affine channel in detectron instead of batch norm layer and i want to freeze this layer. 
In ResNet.py file they use stopgradient but i think it freeze all previous layers.
Is that correct to use stopgradient to freeze a certain one layer?

2. 
I think there is no freeze function for affine layer in resnet
(freeze_at config freezes only chosen blocks not all affine layers)
Is that okay to reproduce the results? Because affine layer is not batch norm layer.
Is there anything freezes affine layer that i missed?

**Comment by youngwanLEE** [comment]
That's my problem, too.

I have tried to replace ResNet backbone with my custom backbone networks that was pretrained in caffe and had batchnorm+scale layer. 

So ,firstly, I converted .caffemodel to .pkl files. 

second, define my backbone networks as like [ResNet.py](https://github.com/facebookresearch/Detectron/blob/master/detectron/modeling/ResNet.py). 

When I define my backbone network that have bachnorm layer, as like ResNet.py, I used AffineChannel layer instead of batchnorm.

My questions are here, 

1. 
How to freeze batchnrom parameter? Faster R-CNN paper mentioned that batchnorm parameters are freezed due to small mini-batch size.

In the [ResNet.py](https://github.com/facebookresearch/Detectron/blob/master/detectron/modeling/ResNet.py#L110), model.StopGradient(s, s) ,I guess, freezes all parameters not only batchnorm parameters(the scale and bias in AffineChannel).  

2. if not freezed batchnorm parameters,  Don't I need to freeze batchnorm parameters?

**Comment by zeenolife** [comment]
There is a method `TrainableParams()` in `DetectionModelHelper` class in `detectron/modeling/detector.py` which returns all trainable params. It depends on `self.do_not_update_params` list, so you can actually somehow append your param names to it manually and that would effectively freeze the weight. That might help.

**Comment by lsrock1** [comment]
@zeenolife thanks!

**Comment by wytcsuch** [comment]
Hi,I wonder to know the effect of the 'inplace‘ in AffineChannel，What's the impact of setting it to True or false? @lsrock1   @youngwanLEE
